deps/rabbit/src/rabbit_queue_member_eval.erl

```diff
@@ -0,0 +1,261 @@
+
+eval_quorum_members(_ExpectedNodes, _Running, [], _TargetSize, Result) ->
+    Result;
+eval_quorum_members(ExpectedNodes, Running, [Q | LocalLeaders], TargetSize, OldResult) ->
+    {ok, Members, {_, LeaderNode}} = ra:members(amqqueue:get_pid(Q)),
+
+    Result =
+    maybe
+        %% Check if Leader is indeed this node
+        LeaderNode ?= node(),
+        %% And that this not is not in maintenance mode
+        true ?= not rabbit_maintenance:is_being_drained_local_read(node()),
+        MemberNodes = [Node || {_, Node} <- Members],
+        Remove = MemberNodes -- ExpectedNodes,
+        case Remove of
+            [] ->
+                add_member(Q, Running, MemberNodes, get_target_size(Q, TargetSize));
+            _ ->
+                remove_members(Q, Remove)
+        end
+    else
+        _ ->
+            noop
+    end,
+    eval_quorum_members(ExpectedNodes, Running, LocalLeaders, TargetSize,
+                        update_result(OldResult, Result)).
+
+add_member(Q, Running, MemberNodes, TargetSize) ->
+    %% Filter out any new nodes under maintenance
+    New = rabbit_maintenance:filter_out_drained_nodes_local_read(Running -- MemberNodes),
+    case should_add_node(MemberNodes, New, TargetSize) of
+        true ->
+            %% In the future, sort the list of new nodes based on load,
+            %% availability zones etc
+            QueueCount = rabbit_amqqueue:count(),
+            QueueCountStartRandom = application:get_env(rabbit, queue_count_start_random_selection,
+                                                        ?QUEUE_COUNT_START_RANDOM_SELECTION),
+            Node = select_node(New, QueueCount, QueueCountStartRandom),
+            QName = amqqueue:get_name(Q),
+            case rabbit_quorum_queue:add_member(Q, Node) of
+                ok ->
+                    rabbit_log:debug(
+                      "Added node ~ts as a member to ~ts as "
+                      "the queues target group size(#~w) is not met and "
+                      "there are enough new nodes(#~w) in the cluster",
+                      [Node, rabbit_misc:rs(QName), TargetSize, length(New)]);
+                {error, Err} ->
+                    rabbit_log:warning(
+                      "~ts: failed to add member (replica) on node ~w, error: ~w",
+                      [rabbit_misc:rs(QName), Node, Err])
+            end,
+            ok;
+        false ->
+            noop
+    end.
+
+should_add_node(MemberNodes, New, TargetSize) ->
+    CurrentSize = length(MemberNodes),
+    NumberOfNewNodes = length(New),
+    maybe
+        true ?= NumberOfNewNodes > 0, %% There are new nodes to grow to
+        true ?= CurrentSize < TargetSize, %% Target size not reached
+        true ?= is_even(CurrentSize) orelse NumberOfNewNodes > 1, %% Enough nodes to grow to odd member size
+        true ?= rabbit_nodes:is_running(lists:delete(node(), MemberNodes))
+    end.
+
+is_even(N) ->
```

Comment:
This belongs to `rabbit_misc` (even though we try to not add things there for its size).